### PR TITLE
pluggable producer/consumer factory

### PR DESCRIPTION
**Pull request by ianynchen**
Current implementation only creates default KafkaConsumer/KafkaProducer, thus making tracing such as zipkin impossible. Provide capability to allow user to provide a custom ConsumerFactory/ProducerFactory, thus allowing the creation of custom Consumer/Producer

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@ianynchen Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=67)!

[Click here](https://cla.pivotal.io/sync/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=67) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@ianynchen Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/reactor?repositoryId=reactor/reactor-kafka&pullRequestId=67)!

**Comment by Haardt** [comment]
If we have the factory interfaces we can use the instrumentation from brave (tracing). Thumbs up :)!
https://github.com/openzipkin/brave/tree/master/instrumentation/kafka-clients#processing-messages
Maybe spring sleuth will use it too?

**Comment by ianynchen** [comment]
@OlegDokuka anything we need to discuss, or you need me to clarify?

**Comment by tran302** [comment]
Is there any plan on when this will be reviewed and released?

**Comment by ianynchen** [comment]
Could someone please take a look at this PR? @OlegDokuka ?

**Comment by ianynchen** [comment]
I'm a bit disappointed. @OlegDokuka are we going to discuss it; discard it or what? This was submitted almost 4 months ago, in the hope that I could use backpressure on my Kafka enabled app while being able to keep my tracing functionalities. But still open without even a word? And during this 4 months time, there was like only 2 PRs merged...

File: src/main/java/reactor/kafka/receiver/internals/DefaultConsumerFactory.java [hunk 1]
```diff
@@ -18,15 +18,17 @@
 import org.apache.kafka.clients.consumer.Consumer;
 import org.apache.kafka.clients.consumer.KafkaConsumer;
 
+import reactor.kafka.receiver.ConsumerFactory;
 import reactor.kafka.receiver.ReceiverOptions;
 
-public class ConsumerFactory {
+public class DefaultConsumerFactory implements ConsumerFactory {
```
Comment (bsideup) [inline src/main/java/reactor/kafka/receiver/internals/DefaultConsumerFactory.java]: ⚠️ API breaking change

File: .gitignore [hunk 2]
```diff
@@ -8,4 +8,4 @@ bin/
 .idea/
 .DS_Store
 *.iml
-
+.idea/**/*
```
Comment (bsideup) [inline .gitignore]: Isn't it already covered with `.idea/`?

**Comment by OlegDokuka** [comment]
@ianynchen Sorry for the late answer, I did not have time to take care of that. 

I don't see a need to introduce a new interface right now since the same can be achieved by extending an existing class. In order to keep builder API, we can keep

```
static <K, V> KafkaReceiver<K, V> create(ConsumerFactory factory, ReceiverOptions<K, V> options) {
        return new DefaultKafkaReceiver<>(factory, options);
```

above improvement without breaking the whole structure.

We are planning to improve an existing API in the 2.x version, which will come soon.


Again, I apologize for so late answer

File: src/main/java/reactor/kafka/sender/ProducerFactory.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package reactor.kafka.sender;
+
+import org.apache.kafka.clients.producer.Producer;
+
+public interface ProducerFactory {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/sender/ProducerFactory.java]: Brakes existing API

File: src/main/java/reactor/kafka/sender/internals/DefaultProducerFactory.java [hunk 4]
```diff
@@ -18,15 +18,17 @@
 import org.apache.kafka.clients.producer.KafkaProducer;
 import org.apache.kafka.clients.producer.Producer;
 
+import reactor.kafka.sender.ProducerFactory;
 import reactor.kafka.sender.SenderOptions;
 
-public class ProducerFactory {
+public class DefaultProducerFactory implements ProducerFactory {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/sender/internals/DefaultProducerFactory.java]: Brakes existing API

File: src/main/java/reactor/kafka/sender/KafkaSender.java [hunk 5]
```diff
@@ -44,8 +44,19 @@
...
+    }
+
+    /**
+     * Creates a Kafka sender that appends records to Kafka topic partitions.
+     * @param factory Custom KafkaProducer factory.
+     * @param options Configuration options of this sender. Changes made to the options
+     *        after the sender is created will not be used by the sender.
+     * @return new instance of Kafka sender
+     */
+    static <K, V> KafkaSender<K, V> create(ProducerFactory factory, SenderOptions<K, V> options) {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/sender/KafkaSender.java]: Let's just keep that method without an introduction of new interfaces

File: src/main/java/reactor/kafka/receiver/KafkaReceiver.java [hunk 6]
```diff
@@ -46,8 +46,25 @@
...
+     * @param factory Custom consumer factory to create a consumer other than the
+     *        default KafkaConsumer. One can use this to wrap a brave.TracingConsumer
+     *        for tracing.
+     * @param options Configuration options of this receiver. Changes made to the options
+     *        after the receiver is created will not be used by the receiver.
+     *        A subscription using group management or a manual assignment of topic partitions
+     *        must be set on the options instance prior to creating this receiver.
+     * @return new receiver instance
+     */
+    static <K, V> KafkaReceiver<K, V> create(ConsumerFactory factory, ReceiverOptions<K, V> options) {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/receiver/KafkaReceiver.java]: Let's just keep this method without an introduction of new interfaces

**Comment by ianynchen** [comment]
@OlegDokuka thank for the update. i'll make the changes later

**Comment by ianynchen** [comment]
@OlegDokuka made the change you requested, now only two methods allowing KafkaReceiver and KafkaSender to be created with a user specified factory

File: src/main/java/reactor/kafka/receiver/KafkaReceiver.java (hunk 6 above)
Comment (ianynchen) [reply src/main/java/reactor/kafka/receiver/KafkaReceiver.java]: done

File: src/main/java/reactor/kafka/sender/KafkaSender.java (hunk 5 above)
Comment (ianynchen) [reply src/main/java/reactor/kafka/sender/KafkaSender.java]: done

**Comment by ianynchen** [comment]
BTW, when is 2.0 scheduled to be released? And if it is to be released soon, will it provide similar changes to this PR and allowing one to create sender/receiver with a custom factory object? Thanks. @OlegDokuka

File: src/main/java/reactor/kafka/receiver/KafkaReceiver.java [hunk 7]
```diff
@@ -46,10 +46,24 @@
      *        must be set on the options instance prior to creating this receiver.
      * @return new receiver instance
      */
-    public static <K, V> KafkaReceiver<K, V> create(ReceiverOptions<K, V> options) {
+    static <K, V> KafkaReceiver<K, V> create(ReceiverOptions<K, V> options) {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/receiver/KafkaReceiver.java]: Why did we change that to package-private? should not it be public?

File: src/main/java/reactor/kafka/sender/KafkaSender.java [hunk 8]
```diff
@@ -44,10 +44,22 @@
      *        after the sender is created will not be used by the sender.
      * @return new instance of Kafka sender
      */
-    public static <K, V> KafkaSender<K, V> create(SenderOptions<K, V> options) {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/sender/KafkaSender.java]: same

File: src/main/java/reactor/kafka/sender/KafkaSender.java [hunk 9]
```diff
@@ -44,10 +44,22 @@
...
 
+    /**
+     * Creates a Kafka sender that appends records to Kafka topic partitions.
+     *
+     * @param factory A custom producer factory other than the default.
+     * @param options Configuration options of this sender. Changes made to the options
+     *        after the sender is created will not be used by the sender.
+     * @return new instance of Kafka sender
+     */
+    static <K, V> KafkaSender<K, V> create(ProducerFactory factory, SenderOptions<K, V> options) {
```
Comment (OlegDokuka) [inline src/main/java/reactor/kafka/sender/KafkaSender.java]: guess it should be public

File: src/main/java/reactor/kafka/receiver/KafkaReceiver.java (hunk 7 above)
Comment (ianynchen) [reply src/main/java/reactor/kafka/receiver/KafkaReceiver.java]: they are by default public. so didn't change them to package-private

File: src/main/java/reactor/kafka/sender/KafkaSender.java (hunk 9 above)
Comment (ianynchen) [reply src/main/java/reactor/kafka/sender/KafkaSender.java]: same as above

Comment (ianynchen) [reply src/main/java/reactor/kafka/sender/KafkaSender.java]: should be public, or at least in java 8. have they changed the rules in java 9 or later?

Comment (OlegDokuka) [reply src/main/java/reactor/kafka/sender/KafkaSender.java]: oh, my bad. I thought it is class

**Pull request merged**